lua/gatekeeper/staticlib.lua

```diff
@@ -74,6 +82,44 @@ function alloc_an_lcore (numa_table)
 	return lcore_t[1]
 end
 
+function merge_arrays (a, b)
+	for i, v in ipairs(b) do
+		table.insert(a, v)
+	end
+	return a
+end
+
+function alloc_lcores_evenly_from_all_numa_nodes (numa_table, n)
+	local num_numa_nodes = get_num_numa_nodes (numa_table)
+	local q = n / num_numa_nodes
+	local r = n % num_numa_nodes
+	local i = 0
+	local res = {}
+	for numa, lcores in all_ipairs(numa_table) do
+		if i < r then
+			if #lcores >= q + 1 then
+				local a1, a2 = split_array(lcores, q + 1)
+				numa_table[numa] = a2
+				numa_table["__net_conf"].numa_used[numa] = true
+				res = merge_arrays(res, a1)
+			else
+				return nil
```

Comment:
Instead of returning `nil` here, we should raise an error because lcores may have been already allocated in previous loop iterations and are not being returned. In order words, this function is not recovering the state if something goes wrong.